passkey-authenticator/src/user_validation.rs

```diff
@@ -1,18 +1,38 @@
+use passkey_types::{ctap2::Ctap2Error, Passkey};
+
 #[cfg(doc)]
 use crate::Authenticator;
 
+/// The result of a user validation check.
+#[derive(Clone, Copy, PartialEq)]
+pub struct UserCheck {
+    /// Indicates whether the user was present.
+    pub presence: bool,
+
+    /// Indicates whether the user was verified.
+    pub verification: bool,
+}
+
 /// Pluggable trait for the [`Authenticator`] to do user interaction and verification.
-#[async_trait::async_trait]
-#[cfg_attr(any(test, feature = "testable"), mockall::automock)]
+#[cfg_attr(any(test, feature = "testable"), mockall::automock(type PasskeyItem = Passkey;))]
 #[cfg_attr(any(test, feature = "testable"), allow(clippy::unused_async))] // Generated by the `mockall` macro.
+#[async_trait::async_trait]
 pub trait UserValidationMethod {
-    /// Check for user verification user a user gesture and confirmation if this operation passes
-    /// both the user presence and user verification flags will be enabled
-    async fn check_user_verification(&self) -> bool;
+    /// The type of the passkey item that can be used to display additional information about the operation to the user.
+    type PasskeyItem: TryInto<Passkey> + Send;
 
-    /// Used when only the user's presence is required, not their verification. For that see
-    /// `check_user_verification`. This will capture the user's consent to the operation.
-    async fn check_user_presence(&self) -> bool;
+    /// Check for the user's presence and obtain consent for the operation. The operation may
+    /// also require the user to be verified.
+    ///
+    /// * `crdential` - Can be used to display additional information about the operation to the user.
+    /// * `presence` - Indicates whether the user's presence is required.
+    /// * `verification` - Indicates whether the user should be verified.
+    async fn check_user(
+        &self,
+        credential: Option<Self::PasskeyItem>,
+        presence: bool,
+        verification: bool,
+    ) -> Result<UserCheck, Ctap2Error>;
```

Comment:
**Suggestion(1 of 4):** This function does not need to own the passkey item passed to it. It can take the item as a reference, like below. Unfortunately Mockall doesn't work that well with references and needs the explicit lifetime.

```suggestion
    async fn check_user<'a>(
        &self,
        credential: Option<&'a Self::PasskeyItem>,
        presence: bool,
        verification: bool,
    ) -> Result<UserCheck, Ctap2Error>;
```